<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="author" content="Thibault Robin">
    <title>Cellosaurus STR Similarity Search Tool - Help</title>
    <link href="css/style.css" rel="stylesheet" type="text/css">
    <link rel="icon" href="https://web.expasy.org/favicon.ico" >
</head>
<body>
    <header id="sib_header_small">
        <div id="sib_expasy_logo" property="schema:isPartOf" typeof="schema:CreativeWork">
            <a href="https://sib.swiss/" id="sib_link"></a>
            <a property="schema:url" href="https://www.expasy.org" id="expasy_link"></a>
        </div>
        <h1><a href="https://web.expasy.org/cellosaurus">Cellosaurus</a></h1>
        <div id="sib_header_nav">
            <a property="" href="https://web.expasy.org/cellosaurus">Home</a>
            &nbsp;|&nbsp;
            <a style="font-weight: bold; color: rgb(204,0,0) !important;" href="/contact">Contact</a>
        </div>
    </header>
    <main class="page">
        <div class="page-header unselectable"><p>Help</p></div>
        <div class="page-body">
            <div class="c1" id="0">Table of contents</div>
            <div class="toc">
                <p><a class="toc c1" href="#1">1. Parameters</a></p>
                <p><a class="toc c2" href="#1.1">1.1 Markers</a></p>
                <p><a class="toc c2" href="#1.2">1.2 Scoring</a></p>
                <p><a class="toc c2" href="#1.3">1.3 Filters</a></p>
                <p><a class="toc c1" href="#2">2. Input File</a></p>
                <p><a class="toc c2" href="#2.1">2.1 Formats</a></p>
                <p><a class="toc c2" href="#2.2">2.2 Compatibility</a></p>
                <p><a class="toc c1" href="#3">3. Results</a></p>
                <p><a class="toc c2" href="#3.1">3.1 Problematic cell lines</a></p>
                <p><a class="toc c2" href="#3.2">3.2 Conflicted markers</a></p>
                <p><a class="toc c2" href="#3.3">3.3 Color indicator</a></p>
                <p><a class="toc c2" href="#3.4">3.4 Nº markers</a></p>
                <p><a class="toc c1" href="#4">4. REST API</a></p>
                <p><a class="toc c2" href="#4.1">4.1 Single entry mode query</a></p>
                <p><a class="toc c2" href="#4.2">4.2 Batch mode query</a></p>
            </div>
            <div class="c1" id="1">1. Parameters</div>
            <div class="c2" id="1.1">1.1 Markers</div>
            <div class="s3">
                <p>
                    Allele data is inputed in the form of values separated by commas. The values are restricted to <span class="t">X</span> and <span class="t">Y</span> (case insensitive)
                    for Amelogenin and numbers for STR markers. The period <span class="t">.</span> character is also allowed in order to describe variant alleles (for example <span class="t">9.3</span>).
                    Note that allele data is whitespace insensitive.
                </p>
            </div>
            <div class="c3" id="1.1.1">1.1.1 Number of alleles</div>
            <p>There is no limit to the number of alleles that can be submitted for a given marker.</p>
            <div class="c3" id="1.1.2">1.1.2 Order of alleles</div>
            <p>The order in which the alleles are entered does not matter. For example submitting <span class="t">14,19.3</span> or <span class="t">19.3,14</span> are equivalent.</p>
            <div class="c3" id="1.1.3">1.1.3 Homozygous loci</div>
            <p>The number of times a homozygous allele is entered does not matter. For example submitting <span class="t">12</span> or <span class="t">12,12</span> are equivalent.
                Homozygous alleles will be counted as one for the score computation.
            </p>
            <div class="c2" id="1.2">1.2 Scoring</div>
            <div class="c3" id="1.2.1">1.2.1 Algorithms</div>
            <p>Three algorithms are provided for the score computation:</p>
            <p><b>Tanabe</b></p>
            <p><i>Also know as Sørensen‐Dice coefficient</i></p>
            <img src='images/equations/tanabe.svg' alt='' height='34'>
            <p>
                DOI=<a class='ab' href='https://www.doi.org/10.11418/jtca1981.18.4_329' target='_blank'>10.11418/jtca1981.18.4_329</a><br>
                PubMed=<a class='ab' href='https://www.ncbi.nlm.nih.gov/pubmed/23136038' target='_blank'>2313603</a>
            </p>
            <p><b>Masters (vs. query)</b></p>
            <p><i>Default version of the Masters algorithm</i></p>
            <img src='images/equations/masters1.svg' alt='' height='34'>
            <p>
                PubMed=<a class='ab' href='https://www.ncbi.nlm.nih.gov/pubmed/11416159' target='_blank'>11416159</a><br>
                PubMed=<a class='ab' href='https://www.ncbi.nlm.nih.gov/pubmed/23136038' target='_blank'>2313603</a>
            </p>
            <p><b>Masters (vs. reference)</b></p>
            <p><i>Modified version of the Masters algorithm</i></p>
            <img src='images/equations/masters2.svg' alt='' height='32'>
            <p>
                PubMed=<a class='ab' href='https://www.ncbi.nlm.nih.gov/pubmed/11416159' target='_blank'>11416159</a><br>
                PubMed=<a class='ab' href='https://www.ncbi.nlm.nih.gov/pubmed/23136038' target='_blank'>2313603</a>
            </p>
            <p>While the Tanabe algorithm is nowadays the recommended algorithm for STR profiling, the Masters algorithms remain useful when trying to figure out the identity of a contaminating cell line.</p>
            <div class="c3" id="1.2.2">1.2.2 Modes</div>
            <p>Three modes are provided for the score computation to handle cases in which allele data is missing for the query or the reference:</p>
            <p><b>Non-empty markers</b></p>
            <p>The score is computed on the markers that have allele data for both the query and the reference.</p>
            <p><b>Query markers</b></p>
            <p>The score is computed on the markers for which the user inputted allele data even if the reference is missing allele data.</p>
            <p><b>References markers</b></p>
            <p>The score is computed on the markers for which reference allele data is available even if the query is missing allele data.</p>
            <img src="images/help/modes.png" alt="Modes"><br>
            <label class="l">Comparison of the three scoring modes. The colored lines indicate the markers that are included in the score computation.</label>
            <div class="c3" id="1.2.3">1.2.3 Amelogenin</div>
            <p>By default, Amelogenin is only used for gender confirmation and is consequently not included in the score computation.
                But you can decide to include it in the scoring.</p>
            <div class="c2" id="1.3">1.3 Filters</div>
            <div class="c3" id="1.3.1">1.3.1 Score filter</div>
            <p>
                Filter defining the minimum score to be reported. Note that in the case of <a href="#3.3">conflicted cell lines</a>, the <span class="t">Best</span> and
                <span class="t">Worst</span> versions are processed as pairs and only the best score is affected by the threshold. Consequently, some <span class="t">Worst</span> cases with
                a score below the threshold can still be present in the results.
            </p>
            <div class="c3" id="1.3.2">1.3.2 Max results</div>
            <p>Filter defining the maximum number of results to be reported.</p>
            <div class="c1" id="2">2. Input File</div>
            <div class="s2">
                <p>
                    Using the <span class="t">Load File</span> button from the user interface, it is possible to directly import STR profile data from a table file. Both mono and multi-samples
                    files are supported. The functionality can be used to perform a similarity search on several samples at a time or to load quickly and reliably the marker data of a sample
                    into the user interface.
                </p>
            </div>
            <div class="c2" id="2.1">2.1 Formats</div>
            <div class="s3">
                <p>The table file can be formatted either as an Excel file (<span class="t">.xls</span> or <span class="t">.xlsx</span> extension) or as a comma separated file
                    (<span class="t">.csv</span> or <span class="t">.txt</span> extension). By default, the tool will assume that each row (except from the header) is a distinct sample.
                    Note that a column named "Sample Name", "Name" or "Sample" is required and each submitted sample needs to have a corresponding value.
                    The ordering of the marker columns is not important.
                    The name of the markers need to be indicated correctly. For Amelogenin, the program recognizes "Amel" and "AM" as valid names.
                </p>
                <table>
                    <tr style="background-color:#ddd"><td>Sample Name</td><td>Amelogenin</td><td>CSF1PO</td><td>D5S818</td><td>D7S820</td><td>D13S317</td><td>D16S539</td><td>TH01</td><td>TPOX</td><td>vWA</td></tr>
                    <tr><td>BICR 16</td><td>X</td><td>12</td><td>13</td><td>10</td><td>11</td><td>12</td><td>6,9</td><td>8,11</td><td>17,19</td></tr>
                    <tr><td>ND31618</td><td>X</td><td>12,13</td><td>10</td><td>8,10</td><td>11,12</td><td>11</td><td>7,9</td><td>8,9</td><td>16</td></tr>
                    <tr><td>Lu-138</td><td>X</td><td>11,12</td><td>10,12</td><td>10,11</td><td>8,11</td><td>11</td><td>7,9</td><td>8,11</td><td>17,18</td></tr>
                    <tr><td>NCI-H125</td><td>X</td><td>7</td><td>10</td><td>10</td><td>11</td><td>5,12</td><td>7</td><td>8,9</td><td>17</td></tr>
                    <tr><td>GK-5</td><td>X,Y</td><td>10,11</td><td>10,12</td><td>10</td><td>12</td><td>11</td><td>9,9.3</td><td>8</td><td>16,19</td></tr>
                </table>
                <label class="l">Example of a properly formatted table</label>
            </div>
            <div class="c2" id="2.2">2.2 Compatibility</div>
            <div class="s3">
                <p>In addition to the format standard described previously, the tool was also made compatible with the following software:</p>
                <ul>
                    <li class="s3"><b>GeneMapper ID-X</b></li>
                </ul>
            </div>
            <div class="c1" id="3">3. Results</div>
            <div class="c2" id="3.1">3.1 Problematic cell lines</div>
            <div class="s3">
                <p>Problematic cell lines, which can be the result of either a contamination or a misidentification, have their accession number displayed in <span style="color:#ee4046;font-weight: bold">red</span> in the result table.
                    When hovering over them, a tooltip displays the relative Cellosaurus information concerning their problematic status.</p>
                <img src="images/help/problematic.png" alt="Problematic cell line"><br>
            </div>
            <div class="c2" id="3.2">3.2 Conflicted markers</div>
            <div class="s3">
                <p>In some cases, sources disagree on the allele values of a given marker. These cases are underlined in the result table and the corresponding sources are indicated in a tooltip when
                    hovering over them. When a cell line possesses conflicted markers, the different combinations of alleles are computed and the <span class="t">Best</span> and <span class="t">Worst</span>
                    reported as distinct rows with an additional label after their accession number.</p>
                <p>For instance, the Loucy cell line (<a href="https://web.expasy.org/cellosaurus/CVCL_1380">CVCL_1380</a>) has a conflict for the <span class="t">D5S818</span> STR marker.
                    DSMZ reports <span class="t">11,12</span> while ATCC and Cosmic-CLP report <span class="t">12</span>:</p>
                <div style="display: flex;">
                    <img style="height: 100%;" src="images/help/conflicts2.png" alt="Conflicted cell line"><br>
                    <img src="images/help/conflicts3.png" alt="Conflicted cell line"><br>
                </div>
                <p>As a result, the score is computed on two different STR profiles: one with <span class="t">11,12</span> and one with <span class="t">12</span>. Based on the scoring results,
                    the profile with <span class="t">11,12</span> as allele value is defined as the <span class="t">Best</span> profile since its a better match for the query.
                    The second profile is consquently defined as the <span class="t">Worst</span> profile:
                </p>
                <img src="images/help/conflicts1.png" alt="Conflicted cell line"><br>
                <p>
                    In more complex cases, a STR marker can have more than two conflicting sources or several STR marker can be conflicted. While the number of generated profile combination is
                    larger, the base principle remains the same. The score is computed for all generated profiles and only the <span class="t">Best</span> and <span class="t">Worst</span> are reported in the results.
                </p>
            </div>
            <div class="c2" id="3.3">3.3 Color indicator</div>
            <div class="s3">
                <p>
                    The color indicator displayed at the left side of each row in the result table allows to easily detect the cell lines that are related. Related cell lines have a
                    <span style="color:#03A45E;font-weight: bold">green</span> indicator while unrelated cell lines have a <span style="color:#ee4046;font-weight: bold">red</span> indicator. The
                    <span style="color:#f59c24;font-weight: bold">orange</span> indicator describes mixed results where related cell lines that have high mutation rates can be found.
                </p>
                <p>The color thresholds based on the score are defined as follows:</p>
                <div style="float: left;margin-right: 25px;">
                    <div><b>Tanabe algorithm</b></div>
                    <table class="cind">
                        <tr class="b1"><td><span class="t">score</span> >= 90</td></tr>
                        <tr class="b3"><td>90 > <span class="t">score</span> >= 80</td></tr>
                        <tr class="b2"><td><span class="t">score</span> < 80</td></tr>
                    </table>
                </div>
                <div>
                    <div><b>Masters algorithms</b></div>
                    <table class="cind">
                        <tr class="b1"><td><span class="t">score</span> >= 80</td></tr>
                        <tr class="b3"><td>80 > <span class="t">score</span> >= 60</td></tr>
                        <tr class="b2"><td><span class="t">score</span> < 60</td></tr>
                    </table>
                </div>
            </div>
            <div class="c2" id="3.4">3.4 Nº markers</div>
            <div class="s3">
                <p>
                    The <span class="t">Nº Markers</span> column reports the number of STR markers that were used in the score computation. It is influenced by the
                    <a href="#1.2.2">scoring mode</a> selected and the inclusion of <a href="#1.2.3">Amelogenin</a> in the score computation. Note that a minimum of eight STR markers is
                    recommended for accurate results.
                </p>
            </div>
            <div class="c1" id="4">4. REST API</div>
            <div class="s2">
                <p>
                    The Cellosaurus STR Similarity Search Tool REST API allows to easily search STR profiles without having to go through the user interface. Two main modes are available:
                    <a href="#4.1">Single entry mode query</a> to search a single sample and <a href="#4.2">Batch mode query</a> to search more than one sample at a time.
                </p>
            </div>
            <div class="c2" id="4.1">4.1 Single entry mode query</div>
            <div></div>
            <div class="c3">URL</div>
            <p style="color:#107dac;">http://129.194.71.205:8080/str-sst/api/query</p>
            <div class="c3">HTTP methods</div>
            <p><span class="get">GET</span>, <span class="post">POST</span><br></p>
            <div class="c3"><span class="post">POST</span> consumes</div>
            <p><span class="t">application/json</span></p>
            <div class="c3"><span class="get">GET</span> & <span class="post">POST</span> produces</div>
            <p><span class="t">application/json</span>&nbsp;&nbsp;<span class="t">text/csv</span></p>
            <div class="c3"><span class="get">GET</span> & <span class="post">POST</span> parameters</div>
            <p>For the <span class="get">GET</span> method, the parameters are added as URL variables after a terminal <span class="t">?</span> character in the URL and separated by <span class="t">&</span> characters. For the <span class="post">POST</span> method, the parameters are inserted in a JSON object as properties.</p>
            <p><i>A more in depth description of the parameters can be found in the <a href="#1">Parameters</a> section. Note that all parameters both keys and values are case insensitive.</i></p>
            <table class="params">
                <tr style="background-color:#ddd"><td><b>"marker name"</b></td><td>Description: marker alleles separated by commas</td></tr>
                <tr><td></td><td>Type: <span class="t">string</span></td></tr>
                <tr style="background-color:#ddd"><td><b>algorithm</b></td><td>Description: scoring algorithm</td></tr>
                <tr><td></td><td>Type: <span class="t">integer</span></td></tr>
                <tr><td></td><td>Choices: <span class="t">1</span> for <a href="#1.3.1">Tanabe</a>, <span class="t">2</span> for <a href="#1.3.1">Masters (vs. query)</a> and <span class="t">3</span> for <a href="#1.3.1">Masters (vs. reference)</a></td></tr>
                <tr><td></td><td>Default: <span class="t">1</span></td></tr>
                <tr style="background-color:#ddd"><td><b>scoringMode</b></td><td>Description: scoring mode </td></tr>
                <tr><td></td><td>Type: <span class="t">integer</span></td></tr>
                <tr><td></td><td>Choices: <span class="t">1</span> for <a href="#1.3.2">Non-empty markers</a>, <span class="t">2</span> for <a href="#1.3.2">Query markers</a> and <span class="t">3</span> for <a href="#1.3.2">Reference markers</a></td></tr>
                <tr><td></td><td>Default: <span class="t">1</span></td></tr>
                <tr style="background-color:#ddd"><td><b>includeAmelogenin</b></td><td>Description: inclusion of Amelogenin in the score computation</td></tr>
                <tr><td></td><td>Type: <span class="t">boolean</span></td></tr>
                <tr><td></td><td>Default: <span class="t">false</span></td></tr>
                <tr style="background-color:#ddd"><td><b>scoreFilter</b></td><td>Description: minimum score</td></tr>
                <tr><td></td><td>Type: <span class="t">integer</span></td></tr>
                <tr><td></td><td>Default: <span class="t">60</span></td></tr>
                <tr style="background-color:#ddd"><td><b>maxResults</b></td><td>Description: maximum number of results</td></tr>
                <tr><td></td><td>Type: <span class="t">integer</span></td></tr>
                <tr><td></td><td>Default: <span class="t">200</span></td></tr>
                <tr style="background-color:#ddd"><td><b>description</b></td><td>Description: optional tag describing the query</td></tr>
                <tr><td></td><td>Type: <span class="t">string</span></td></tr>
                <tr style="background-color:#ddd"><td><b>outputFormat</b></td><td>Description: format of the API output</td></tr>
                <tr><td></td><td>Type: <span class="t">string</span></td></tr>
                <tr><td></td><td>Choices: <span class="t">json</span> or <span class="t">csv</span></td></tr>
                <tr><td></td><td>Default: <span class="t">json</span></td></tr>
            </table><br>
            <div class="c3"><span class="get">GET</span> & <span class="post">POST</span> response status code</div>
            <p><span class="t">200</span></p>
            <div class="c3"><span class="get">GET</span> example</div>
            <a style="word-break:break-all" href="http://129.194.71.205:8080/str-sst/api/query?Amelogenin=X&CSF1PO=13,14&D5S818=13&D7S820=8,9&D13S317=12&FGA=24&TH01=8&TPOX=11&vWA=16&algorithm=1&scoringMode=1&scoreFilter=70&includeAmelogenin=false" target="_blank">http://129.194.71.205:8080/str-sst/api/query?Amelogenin=X&CSF1PO=13,14&D5S818=13&D7S820=8,9&D13S317=12&FGA=24&TH01=8&TPOX=11&vWA=16&algorithm=1&scoringMode=1&scoreFilter=70&includeAmelogenin=false</a>
            <div class="c3"><span class="post">POST</span> example</div>
            <p><span class="t">{<br>
                    "Amelogenin": "X",<br>
                    "CSF1PO": "13,14",<br>
                    "D5S818": "13",<br>
                    "D7S820": "8",<br>
                    "D13S317": "12",<br>
                    "FGA": "24",<br>
                    "TH01": "8",<br>
                    "TPOX": "11",<br>
                    "vWA": "16",<br>
                    "algorithm": 1,<br>
                    "scoringMode": 1,<br>
                    "scoreFilter": 70,<br>
                    "includeAmelogenin": false<br>
                    }</span></p>
            <div class="c2" id="4.2">4.2 Batch mode query</div>
            <div class="c3">URL</div>
            <p style="color:#107dac;">http://129.194.71.205:8080/str-sst/api/batch</p>
            <div class="c3">HTTP method</div>
            <p><span class="post">POST</span><br></p>
            <div class="c3"><span class="post">POST</span> consumes</div>
            <p><span class="t">application/json</span></p>
            <div class="c3"><span class="post">POST</span> produces</div>
            <p><span class="t">application/json</span>&nbsp;&nbsp;<span class="t">application/zip</span></p>
            <div class="c3"><span class="post">POST</span> parameters</div>
            <p> The parameters of each sample to be searched are inserted in a JSON Array as distinct elements, as seen in the <a href="#BatchPostExample">Example</a>.
                When the <span class="t">csv</span> output format is selected, the produced <span class="t">csv</span> files are compressed together in a  <span class="t">zip</span> file.
            </p>
            <p><i>A more in depth description of the parameters can be found in the <a href="#1">Parameters</a> section. Note that all parameters, both keys and values, are case insensitive.</i></p>
            <table class="params">
                <tr style="background-color:#ddd"><td><b>"marker name"</b></td><td>Description: marker alleles separated by commas</td></tr>
                <tr><td></td><td>Type: <span class="t">string</span></td></tr>
                <tr style="background-color:#ddd"><td><b>algorithm</b></td><td>Description: scoring algorithm</td></tr>
                <tr><td></td><td>Type: <span class="t">integer</span></td></tr>
                <tr><td></td><td>Choices: <span class="t">1</span> for <a href="#1.3.1">Tanabe</a>, <span class="t">2</span> for <a href="#1.3.1">Masters (vs. query)</a> and <span class="t">3</span> for <a href="#1.3.1">Masters (vs. reference)</a></td></tr>
                <tr><td></td><td>Default: <span class="t">1</span></td></tr>
                <tr style="background-color:#ddd"><td><b>scoringMode</b></td><td>Description: scoring mode </td></tr>
                <tr><td></td><td>Type: <span class="t">integer</span></td></tr>
                <tr><td></td><td>Choices: <span class="t">1</span> for <a href="#1.3.2">Non-empty markers</a>, <span class="t">2</span> for <a href="#1.3.2">Query markers</a> and <span class="t">3</span> for <a href="#1.3.2">Reference markers</a></td></tr>
                <tr><td></td><td>Default: <span class="t">1</span></td></tr>
                <tr style="background-color:#ddd"><td><b>includeAmelogenin</b></td><td>Description: inclusion of Amelogenin in the score computation</td></tr>
                <tr><td></td><td>Type: <span class="t">boolean</span></td></tr>
                <tr><td></td><td>Default: <span class="t">false</span></td></tr>
                <tr style="background-color:#ddd"><td><b>scoreFilter</b></td><td>Description: minimum score</td></tr>
                <tr><td></td><td>Type: <span class="t">integer</span></td></tr>
                <tr><td></td><td>Default: <span class="t">60</span></td></tr>
                <tr style="background-color:#ddd"><td><b>maxResults</b></td><td>Description: maximum number of results</td></tr>
                <tr><td></td><td>Type: <span class="t">integer</span></td></tr>
                <tr><td></td><td>Default: <span class="t">200</span></td></tr>
                <tr style="background-color:#ddd"><td><b>description</b></td><td>Description: optional tag describing the query</td></tr>
                <tr><td></td><td>Type: <span class="t">string</span></td></tr>
                <tr style="background-color:#ddd"><td><b>outputFormat</b></td><td>Description: format of the API output</td></tr>
                <tr><td></td><td>Type: <span class="t">string</span></td></tr>
                <tr><td></td><td>Choices: <span class="t">json</span> or <span class="t">csv</span></td></tr>
                <tr><td></td><td>Default: <span class="t">json</span></td></tr>
            </table><br>
            <div class="c3"><span class="post">POST</span> response status code</div>
            <p><span class="t">200</span></p>
            <div class="c3"><span class="post" id="BatchPostExample">POST</span> example</div>
            <p><span class="t">[{<br>
                    "description": "Example 1",<br>
                    "Amelogenin": "X",<br>
                    "CSF1PO": "13,14",<br>
                    "D5S818": "13",<br>
                    "D7S820": "8",<br>
                    "D13S317": "12",<br>
                    "FGA": "24",<br>
                    "TH01": "8",<br>
                    "TPOX": "11",<br>
                    "vWA": "16",<br>
                    "algorithm": 2,<br>
                    "scoringMode": 1,<br>
                    "scoreFilter": 70,<br>
                    "includeAmelogenin": true<br>
                    },{<br>
                    "description": "Example 2",<br>
                    "Amelogenin": "X, Y",<br>
                    "CSF1PO": "13",<br>
                    "D5S818": "13, 14",<br>
                    "D7S820": "8, 19",<br>
                    "D13S317": "11, 12",<br>
                    "FGA": "24",<br>
                    "TH01": "8",<br>
                    "TPOX": "11",<br>
                    "vWA": "15",<br>
                    "algorithm": 2,<br>
                    "scoringMode": 1,<br>
                    "scoreFilter": 70,<br>
                    "includeAmelogenin": true<br>
                    }]</span></p>
            <div id="back"><a href="index.html">Back</a></div>
        </div>
    </main>
    <footer id="sib_footer">
        <div id="sib_footer_content" property="schema:creator" typeof="schema:NGO">
            <a property="schema:url" href="https://sib.swiss/"><span property="schema:name">SIB Swiss Institute of Bioinformatics</span></a><span>&nbsp;|&nbsp;</span><a href="https://www.expasy.org/disclaimer.html">Disclaimer</a>
            <div id="sib_footer_right">
                <a href="#" id="sib_footer_gototop"><span style="padding-left: 10px">Back to the Top</span></a>
            </div>
        </div>
    </footer>
    <button id="scroller" onclick="scrollUp()">TOP</button>
    <script>
        window.onscroll = function () {
            document.getElementById("scroller").style.display = document.body.scrollTop > 20 || document.documentElement.scrollTop > 20 ? "block" : "none";
        };
        function scrollUp() {
            document.body.scrollIntoView({behavior: 'instant', block: "start", inline: "nearest"});
        }
    </script>
</body>
</html>
